packages/headless/src/controllers/triggers/headless-query-trigger.ts

```diff
@@ -0,0 +1,82 @@
+import {Engine} from '../../app/headless-engine';
+import {ConfigurationSection, TriggerSection} from '../../state/state-sections';
+import {configuration, triggers} from '../../app/reducers';
+import {buildController, Controller} from '../controller/headless-controller';
+import {loadReducerError} from '../../utils/errors';
+import {executeSearch} from '../../features/search/search-actions';
+import {
+  updateQuery,
+  UpdateQueryActionCreatorPayload,
+} from '../../features/query/query-actions';
+import {logTriggerQuery} from '../../features/triggers/trigger-analytics-actions';
+
+/**
+ * The `QueryTrigger` controller handles query triggers.
+ */
+export interface QueryTrigger extends Controller {
+  /**
+   * the state of the `QueryTrigger` controller.
+   */
+  state: QueryTriggerState;
+}
+
+/**
+ * A scoped and simplified part of the headless state that is relevant to the `QueryTrigger` controller.
+ */
+export interface QueryTriggerState {
+  /**
+   * The new query to perform a search with after receiving a query trigger.
+   */
+  query: string | null;
+}
+
+/**
+ * Creates a `QueryTrigger` controller instance.
+ *
+ * @param engine - The headless engine.
+ * @returns A `QueryTrigger` controller instance.
+ * */
+export function buildQueryTrigger(engine: Engine<object>): QueryTrigger {
+  if (!loadQueryTriggerReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const controller = buildController(engine);
+  const {dispatch} = engine;
+
+  const getState = () => engine.state;
+
+  return {
+    ...controller,
+
+    subscribe(listener: () => void) {
+      const strictListener = () => {
+        if (this.state.query) {
+          listener();
+          const updateQueryPayload: UpdateQueryActionCreatorPayload = {
+            q: this.state.query,
+          };
+          dispatch(updateQuery(updateQueryPayload));
+          //maybe logSearchBoxSubmit instead
+          //and dispatch logTriggerQuery  separately?
+          dispatch(executeSearch(logTriggerQuery()));
+        }
+      };
+      strictListener();
+      return engine.subscribe(strictListener);
+    },
+
+    get state() {
+      return {
+        query: getState().triggers.query,
```

Comment:
This controller should expose the original query typed by the user, as well as the corrected query returned by the api.
This would allow a developer to display a message in the UI, explaining to the user that the query has changed from `a` to `b`.